### PR TITLE
fix: support `associated_const_equality` in parser

### LINKED ISSUE #11965: False syntax errors with `associated_const_equality`
The following code works in nightly with `cargo check` but rust-analyzer reports `HasCount<COUNT = {N}>` as containing syntax errors:

```rust
#![feature(associated_const_equality)]

trait HasCount {
    const COUNT: usize;
}

trait Foo {}

impl<T, const N: usize> Foo for T where T: HasCount<COUNT = {N}> {}
```

Using pre-release VSCode extension. The official Rust plugin works fine.

### PR DESCRIPTION
**Pull request by XFFXFF**
This pr fixes #11965.  The parser now allows eq constraints on associated constants.  

I've added tests for `HasCount<Count = {N}>` and `HasCount<Count = 0>`

### CODE REVIEW

File: crates/parser/test_data/parser/ok/0073_associated_const_equality.rs [hunk 1]
```diff
@@ -0,0 +1,3 @@
+impl<T, const N: usize> Foo for T where T: HasCount<COUNT = {N}> {}
+
+impl<T> Foo for T where T: HasCount<COUNT = 0> {}
```
Comment (Veykril) [inline crates/parser/test_data/parser/ok/0073_associated_const_equality.rs:3]: These tests should be written as comments in the source code and then auto-generated by running `cargo test -p syntax`. See the `// test test_name` comments in the parser modules.

Comment (XFFXFF) [reply crates/parser/test_data/parser/ok/0073_associated_const_equality.rs]: I have written these tests as comments and changed them a bit.  
It seems that `cargo test -p syntax` doesn't automatically generate tests files, instead, I manually run `sourcegen_parser_tests`

Comment (Veykril) [reply crates/parser/test_data/parser/ok/0073_associated_const_equality.rs]: Ah my bad, its `cargo test -p parser`

**Comment by XFFXFF** [comment]
Make the CI happy again :). Please take a look @Veykril

**Comment by Veykril** [comment]
Thanks!
@bors r+

**Pull request merged**